metrics/sinks/wavefront/wavefront.go

```diff
@@ -0,0 +1,280 @@
+			} else {
+				//do nothing for now
+				metricValStr = ""
+			}
+			if metricValStr != "" {
+				ts := strconv.FormatInt(batch.Timestamp.Unix(), 10)
+				stype := tags["type"]
+				source := ""
+				if stype == "cluster" {
+					source = this.ClusterName
+				} else if stype == "ns" {
+					source = tags["namespace_name"] + "-ns"
+				} else {
+					source = tags["hostname"]
+				}
+				tagStr := tagsToString(tags)
+				this.SendPoint(this.cleanMetricName(metricName), metricValStr, ts, source, tagStr)
+				metricCounter = metricCounter + 1
+			}
+		}
+		for _, metric := range ms.LabeledMetrics {
+			metricName := this.cleanMetricName(metric.Name)
+			metricValStr := ""
+			if core.ValueInt64 == metric.ValueType {
+				metricValStr = fmt.Sprintf("%d", metric.IntValue)
+			} else if core.ValueFloat == metric.ValueType { // W
+				metricValStr = fmt.Sprintf("%f", metric.FloatValue)
+			} else {
+				//do nothing for now
+				metricValStr = ""
+			}
+			if metricValStr != "" {
+				ts := strconv.FormatInt(batch.Timestamp.Unix(), 10)
+				source := tags["hostname"]
+				tagStr := tagsToString(tags)
+				for labelName, labelValue := range metric.Labels {
+					tagStr += labelName + "=\"" + labelValue + "\" "
+				}
+				metricCounter = metricCounter + 1
+				this.SendPoint(metricName, metricValStr, ts, source, tagStr)
+			}
+		}
+	}
+
+}
+
+func (this *WavefrontSink) ExportData(batch *core.DataBatch) {
+	//make sure we're Connected
+	err := this.Connect()
+	if err != nil {
+		glog.Warning(err)
+	}
+
+	if this.Conn != nil && err == nil {
+		this.Send(batch)
+	}
+}
+
+func (this *WavefrontSink) Connect() error {
+	var err error
+	this.Conn, err = net.DialTimeout("tcp", this.ProxyAddress, time.Second*10)
+	if err != nil {
+		glog.Warningf("Unable to connect to Wavefront proxy at address: %s", this.ProxyAddress)
+		return err
+	} else {
+		glog.Infof("Connected to Wavefront proxy at address: %s", this.ProxyAddress)
+		return nil
+	}
+}
+
+func NewWavefrontSink(uri *url.URL) (core.DataSink, error) {
+
+	storage := &WavefrontSink{
+		ProxyAddress:      uri.Scheme + ":" + uri.Opaque,
+		ClusterName:       "k8s-cluster",
+		Prefix:            "heapster.",
+		IncludeLabels:     false,
+		IncludeContainers: true,
+	}
+
+	vals := uri.Query()
+	if len(vals["clusterName"]) > 0 {
+		storage.ClusterName = vals["clusterName"][0]
+	}
+	if len(vals["prefix"]) > 0 {
+		storage.Prefix = vals["prefix"][0]
+	}
+	if len(vals["includeLabels"]) > 0 {
+		incLabels := false
+		incLabels, err := strconv.ParseBool(vals["includeLabels"][0])
+		if err != nil {
+			return nil, err
+		}
+		storage.IncludeLabels = incLabels
+	}
+	if len(vals["includeContainers"]) > 0 {
+		incContainers := false
+		incContainers, err := strconv.ParseBool(vals["includeContainers"][0])
+		if err != nil {
+			return nil, err
+		}
+		storage.IncludeContainers = incContainers
+	}
+	return storage, nil
+}
+
+func sortedMetricSetKeys(m map[string]*core.MetricSet) []string {
```

Comment:
I don't disagree, other than the maps being of different types. We would still need three separate functions because they have different signatures. We could add a single function that accepts a map[string]interface{} then do type assertion. This may be less performant though.